src/main/java/com/openlattice/data/DataGraph.java

```diff
@@ -28,65 +28,66 @@
 import com.google.common.collect.SetMultimap;
 import com.openlattice.client.serialization.SerializationConstants;
 
-import java.util.Map;
-import java.util.Objects;
-import java.util.Set;
-import java.util.UUID;
+import java.util.*;
 
 /**
  * @author Matthew Tamayo-Rios &lt;matthew@openlattice.com&gt;
  */
 public class DataGraph {
-    private final ListMultimap<UUID, Map<UUID, Set<Object>>> entities;
-    private final ListMultimap<UUID, DataAssociation>        associations;
+    private final Map<UUID, List<Map<UUID, Set<Object>>>> entities;
```

Comment:
What's the reason for these ListMultimap<UUID, Blah> -> Map<UUID, List<Blah >> changes? Do these calls now return a different object that they used to (i.e. including keys with empty values)?

Reply:
To get rid of multimaps in general. I'm not sure about the empty values..etc, will get back after some investigation.. so far none of the DataControllerTest/DataControllerLinkingTest rehearsal tests broke bc of this.

Reply:
DataGraph we only deserialize from JSON (in `DataApi.createEntityAndAssociationData`)
There can be differences in `Map<A,List<B>>` and `ListMultimap<A,B>` deserializations:
* First is, what you said, in map there can be keys with empty values and those will be still there after deserialization. This could be easily solved by filtering empty values + you can introduce empty values in multimaps, that have a collection as a value.
    * Map:`{"a": [1,1], "b": []}` -> `{a=[1,1], b=[]}`
    * Multimap: `{"a": [1,1], "b": []}` -> `{a=[1,1]}`
    * Map:`{"a": [{"a": 1}], "b": []}` -> `{a=[{a=1}], b=[{}]}`
    * Multimap: `{a=[{a=1}], b=[{}]}` -> `{a=[{a=1}], b=[{}]}`
- Second (and bigger issue) is, that if the same key is present multiple times in the JSON, multimap will keep all the values, while map deserialization will only keep the last one:
    * Map:`{"a": [1], "a": [1]}` -> `{a=[1]}`
    * Multimap: `{"a": [1], "a": [1]}` -> `{a=[1,1]}`

Reply:
Is there a way to check how we use this endpoint in APPs? If we can avoid the second issue, there shouldn't be a problem with DataGraph changes.

Reply:
It shouldn't be possible to pass a map from javascript that contains the same key multiple times, so the second case should be fine. And yeah, if this object is only ever passed from the frontend to the backend this should be fine -- the place potential subtle errors would commonly occur when moving away from multimaps is in the case of a pattern like:

```
fun getBlahForObjects(objectIds: Iterable<UUID>): SetMultimap<UUID, Blah> { ... }
```
vs
```
fun getBlahForObjects(objectIds: Iterable<UUID>): Map<UUID, Set<Blah>> { ... }
```

where the frontend passes `[uuid1, uuid2]` where the first endpoint would return
```
{
  <uuid1>: [<blah1>]
}
```
and the second returns
```
{
  <uuid1>: [<blah1>],
  <uuid2>: []
}
```

This could be problematic if the frontend is doing something like:
```
const objectsWithBlah = Object.keys(getBlahForObjects([uuid1, uuid2]))
```
since now uuid2 would be treated as having blah, when it does not actually have blah.

Reply:
Yeah, that can cause problems, if any of the calls on frontend expect non-empty results. I think, if we don't want to investigate which frontend app could possibly break from this, I can change it, so that the main constructor will have multimaps, but store the values as maps after, so we don't deal with multimaps in code, that handles the call, only for deserializing.
```
    @Deprecated(since = "Use other constructor instead.")
    @JsonCreator
    public DataGraph(
            @JsonProperty( SerializationConstants.ENTITIES ) ListMultimap<UUID, Map<UUID, Set<Object>>> entities,
            @JsonProperty( SerializationConstants.ASSOCIATIONS ) ListMultimap<UUID, DataAssociation> associations
    ) {
        this.entities = Multimaps.asMap( entities );
        this.associations = Multimaps.asMap( associations );
    }
```